Gym/DQN/main.swift

```diff
@@ -0,0 +1,278 @@
+    typealias Output = Tensor<Float>
+
+    var l1, l2: Dense<Float>
+
+    init(observationSize: Int, hiddenSize: Int, actionCount: Int) {
+        l1 = Dense<Float>(inputSize: observationSize, outputSize: hiddenSize, activation: relu, weightInitializer: heNormal())
+        l2 = Dense<Float>(inputSize: hiddenSize, outputSize: actionCount, weightInitializer: heNormal())
+    }
+
+    @differentiable
+    func callAsFunction(_ input: Input) -> Output {
+        return input.sequenced(through: l1, l2)
+    }
+}
+
+class Agent {
+    // Q-network
+    var qNet: Net
+    // Target Q-network
+    var targetQNet: Net
+    // Optimizer
+    let optimizer: Adam<Net>
+    // Replay Buffer
+    let replayBuffer: ReplayBuffer
+    // Discount Factor
+    let discount: Float
+
+    init(qNet: Net, targetQNet: Net, optimizer: Adam<Net>, replayBuffer: ReplayBuffer, discount: Float) {
+        self.qNet = qNet
+        self.targetQNet = targetQNet
+        self.optimizer = optimizer
+        self.replayBuffer = replayBuffer
+        self.discount = discount
+    }
+
+    func getAction(state: Tensor<Float>, epsilon: Float) -> Tensor<Int32> {
+        if Float(np.random.uniform()).unwrapped() < epsilon {
+            // print("getAction | state: \(state)")
+            // print("getAction | epsilon: \(epsilon)")
+            let npState = np.random.randint(0, 2, dtype: np.int32)
+            // print("getAction | npState: \(npState)")
+            return Tensor<Int32>(numpy: np.array(npState, dtype: np.int32))!
+        }
+        else {
+            // Neural network input needs to be 2D
+            let tfState = Tensor<Float>(numpy: np.expand_dims(state.makeNumpyArray(), axis: 0))!
+            let qValues = qNet(tfState)
+            let leftQValue = Float(qValues[0][0]).unwrapped()
+            let rightQValue = Float(qValues[0][1]).unwrapped()
+            return leftQValue < rightQValue ? Tensor<Int32>(numpy: np.array(1, dtype: np.int32))! : Tensor<Int32>(numpy: np.array(0, dtype: np.int32))!
+        }
+    }
+
+    func train(batchSize: Int) {
+        // Don't train if replay buffer is too small
+        if replayBuffer.count >= batchSize {
+            // print("train | Start training")
+            let (tfStateBatch, tfActionBatch, tfRewardBatch, tfNextStateBatch) = replayBuffer.sample(batchSize: batchSize)
+
+            // TODO: Find equivalent function of tf.gather_nd in S4TF to parallelize Q-value computation (_Raw.gather_nd does not exist)
+            // Gradient are accumulated since we calculate every element in the batch individually
+            var totalGrad = qNet.zeroTangentVector
+            for i in 0..<batchSize {
+                let 𝛁qNet = gradient(at: qNet) { qNet -> Tensor<Float> in
+
+                    let stateQValueBatch = qNet(tfStateBatch)
+                    let tfAction: Tensor<Int32> = tfActionBatch[i][0]
+                    let action = Int(tfAction.makeNumpyArray()).unwrapped()
```

Comment:
You can use [`Tensor.scalarized()`](https://www.tensorflow.org/swift/api_docs/Structs/Tensor#scalarized) to concisely convert a scalar `Tensor<Scalar>` into a `Scalar`.
```suggestion
                    let action = Int(tfAction.scalarized())
```

Reply:
Amazing! This makes everything so much cleaner.